File: csharp/src/Fory/BFloat16.cs
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

namespace Apache.Fory;

/// <summary>
/// Represents one IEEE 754 bfloat16 value.
/// </summary>
/// <remarks>
/// <para>
/// <see cref="BFloat16" /> is the public Fory carrier for xlang <c>BFLOAT16</c> values. It stores
/// the canonical 16-bit wire representation and converts to and from <see cref="float" /> using
/// round-to-nearest-even semantics.
/// </para>
/// <para>
/// Use <see cref="FromBits" /> and <see cref="ToBits" /> when you need exact bit-preserving
/// control. Use <see cref="FromSingle" /> or the explicit conversion operators when you want
/// numeric conversion.
/// </para>
/// <para>
/// For xlang <c>bfloat16_array</c> payloads, use <c>BFloat16[]</c> or <c>List&lt;BFloat16&gt;</c>.
/// Both carriers map to the packed 16-bit array wire format, so a dedicated list wrapper is not
/// required.
/// </para>
/// </remarks>
public readonly struct BFloat16 : IEquatable<BFloat16>, IComparable<BFloat16>
{
    private const uint Float32ExpMask = 0x7F80_0000u;
    private const uint Float32MantissaMask = 0x007F_FFFFu;
    private const ushort CanonicalNaNBits = 0x7FC0;

    /// <summary>
    /// Gets the canonical quiet NaN value.
    /// </summary>
    public static BFloat16 NaN => FromBits(CanonicalNaNBits);

    /// <summary>
    /// Gets positive infinity.
    /// </summary>
    public static BFloat16 PositiveInfinity => FromBits(0x7F80);

    /// <summary>
    /// Gets negative infinity.
    /// </summary>
    public static BFloat16 NegativeInfinity => FromBits(0xFF80);

    /// <summary>
    /// Gets positive zero.
    /// </summary>
    public static BFloat16 Zero => default;

    /// <summary>
    /// Gets negative zero.
    /// </summary>
    public static BFloat16 NegativeZero => FromBits(0x8000);

    /// <summary>
    /// Gets the raw IEEE 754 bfloat16 bits.
    /// </summary>
    public ushort Bits { get; }

    /// <summary>
    /// Initializes a new instance from the exact wire bits.
    /// </summary>
    /// <param name="bits">Raw bfloat16 bits.</param>
    public BFloat16(ushort bits)
    {
        Bits = bits;
    }

    /// <summary>
    /// Creates a value from exact wire bits.
    /// </summary>
    public static BFloat16 FromBits(ushort bits) => new(bits);

    /// <summary>
    /// Creates a value from a <see cref="float" />.
    /// </summary>
    public static BFloat16 FromSingle(float value) => new(Float32ToBFloat16Bits(value));

    /// <summary>
    /// Returns the exact wire bits.
    /// </summary>
    public ushort ToBits() => Bits;

    /// <summary>
    /// Converts this value to <see cref="float" />.
    /// </summary>
    public float ToSingle() => BFloat16BitsToFloat32(Bits);

    /// <summary>
    /// Converts this value to <see cref="double" />.
    /// </summary>
    public double ToDouble() => ToSingle();

    /// <summary>
    /// Returns whether the value is NaN.
    /// </summary>
    public bool IsNaN => (Bits & 0x7F80) == 0x7F80 && (Bits & 0x007F) != 0;

    /// <summary>
    /// Returns whether the value is infinite.
    /// </summary>
    public bool IsInfinity => (Bits & 0x7F80) == 0x7F80 && (Bits & 0x007F) == 0;

    /// <summary>
    /// Returns whether the value is finite.
    /// </summary>
    public bool IsFinite => (Bits & 0x7F80) != 0x7F80;

    /// <summary>
    /// Returns whether the value is numerically zero, including signed zero.
    /// </summary>
    public bool IsZero => (Bits & 0x7FFF) == 0;

    /// <summary>
    /// Returns whether the sign bit is set.
    /// </summary>
    public bool SignBit => (Bits & 0x8000) != 0;

    /// <summary>
    /// Reinterprets a <see cref="float" /> as bfloat16 bits using round-to-nearest-even.
    /// </summary>
    public static ushort ToBits(float value) => Float32ToBFloat16Bits(value);

    /// <summary>
    /// Reinterprets bfloat16 bits as a <see cref="float" />.
    /// </summary>
    public static float ToSingle(ushort bits) => BFloat16BitsToFloat32(bits);

    /// <summary>
    /// Converts from <see cref="float" /> to <see cref="BFloat16" />.
    /// </summary>
    public static explicit operator BFloat16(float value) => FromSingle(value);

    /// <summary>
    /// Converts from <see cref="double" /> to <see cref="BFloat16" />.
    /// </summary>
    public static explicit operator BFloat16(double value) => FromSingle((float)value);

    /// <summary>
    /// Converts to <see cref="float" />.
    /// </summary>
    public static implicit operator float(BFloat16 value) => value.ToSingle();

    /// <summary>
    /// Converts to <see cref="double" />.
    /// </summary>
    public static implicit operator double(BFloat16 value) => value.ToDouble();

    public int CompareTo(BFloat16 other)
    {
        return ToSingle().CompareTo(other.ToSingle());
    }

    public bool Equals(BFloat16 other)
    {
        return Bits == other.Bits;
    }

    public override bool Equals(object? obj)
    {
        return obj is BFloat16 other && Equals(other);
    }

    public override int GetHashCode()
    {
        return Bits.GetHashCode();
    }

    public override string ToString()
    {
        return ToSingle().ToString(System.Globalization.CultureInfo.InvariantCulture);
    }

    public static bool operator ==(BFloat16 left, BFloat16 right) => left.Equals(right);

    public static bool operator !=(BFloat16 left, BFloat16 right) => !left.Equals(right);

    public static bool operator <(BFloat16 left, BFloat16 right) => left.CompareTo(right) < 0;

    public static bool operator <=(BFloat16 left, BFloat16 right) => left.CompareTo(right) <= 0;

    public static bool operator >(BFloat16 left, BFloat16 right) => left.CompareTo(right) > 0;

    public static bool operator >=(BFloat16 left, BFloat16 right) => left.CompareTo(right) >= 0;

    internal static ushort Float32ToBFloat16Bits(float value)
    {
        uint bits32 = BitConverter.SingleToUInt32Bits(value);
        if ((bits32 & Float32ExpMask) == Float32ExpMask && (bits32 & Float32MantissaMask) != 0)
        {
            return CanonicalNaNBits;
        }

        uint lsb = (bits32 >> 16) & 1u;
        uint rounded = bits32 + 0x7FFFu + lsb;
        return unchecked((ushort)(rounded >> 16));
    }

    internal static float BFloat16BitsToFloat32(ushort bits)
    {
        return BitConverter.UInt32BitsToSingle((uint)bits << 16);
    }
}

File: rust/fory-core/src/types/bfloat16.rs
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

//! IEEE 754 bfloat16 floating-point type.
//!
//! This module provides a `bfloat16` type that stores an IEEE 754 bfloat16
//! payload exactly. The type is a transparent wrapper around `u16` and
//! provides round-to-nearest-even conversion from `f32`, exact expansion back
//! to `f32`, classification helpers, and arithmetic through `f32`.
//!
//! The type is re-exported from `fory_core` as `BFloat16`, and `Vec<bfloat16>`
//! / `Vec<BFloat16>` is the canonical dense carrier for xlang `bfloat16_array`
//! payloads.

use std::cmp::Ordering;
use std::fmt;
use std::hash::{Hash, Hasher};
use std::ops::{Add, Div, Mul, Neg, Sub};

#[repr(transparent)]
#[derive(Copy, Clone, Default)]
#[allow(non_camel_case_types)]
pub struct bfloat16(u16);

const SIGN_MASK: u16 = 0x8000;
const EXP_MASK: u16 = 0x7F80;
const MANTISSA_MASK: u16 = 0x007F;
const INFINITY_BITS: u16 = 0x7F80;
const NEG_INFINITY_BITS: u16 = 0xFF80;
const QUIET_NAN_BITS: u16 = 0x7FC0;

impl bfloat16 {
    #[inline(always)]
    pub const fn from_bits(bits: u16) -> Self {
        Self(bits)
    }

    #[inline(always)]
    pub const fn to_bits(self) -> u16 {
        self.0
    }

    pub const ZERO: Self = Self(0x0000);
    pub const NEG_ZERO: Self = Self(0x8000);
    pub const INFINITY: Self = Self(INFINITY_BITS);
    pub const NEG_INFINITY: Self = Self(NEG_INFINITY_BITS);
    pub const NAN: Self = Self(QUIET_NAN_BITS);
    pub const MAX: Self = Self(0x7F7F);
    pub const MIN_POSITIVE: Self = Self(0x0080);
    pub const MIN_POSITIVE_SUBNORMAL: Self = Self(0x0001);

    #[inline(always)]
    pub fn from_f32(value: f32) -> Self {
        let bits = value.to_bits();
        if (bits & 0x7F80_0000) == 0x7F80_0000 && (bits & 0x007F_FFFF) != 0 {
            return Self(QUIET_NAN_BITS);
        }
        let lsb = (bits >> 16) & 1;
        let rounding_bias = 0x7FFF + lsb;
        Self(((bits + rounding_bias) >> 16) as u16)
    }

    #[inline(always)]
    pub fn to_f32(self) -> f32 {
        f32::from_bits((self.0 as u32) << 16)
    }

    #[inline(always)]
    pub fn is_nan(self) -> bool {
        (self.0 & EXP_MASK) == EXP_MASK && (self.0 & MANTISSA_MASK) != 0
    }

    #[inline(always)]
    pub fn is_infinite(self) -> bool {
        (self.0 & EXP_MASK) == EXP_MASK && (self.0 & MANTISSA_MASK) == 0
    }

    #[inline(always)]
    pub fn is_finite(self) -> bool {
        (self.0 & EXP_MASK) != EXP_MASK
    }

    #[inline(always)]
    pub fn is_zero(self) -> bool {
        (self.0 & !SIGN_MASK) == 0
    }

    #[inline(always)]
    pub fn is_normal(self) -> bool {
        let exp = self.0 & EXP_MASK;
        exp != 0 && exp != EXP_MASK
    }

    #[inline(always)]
    pub fn is_subnormal(self) -> bool {
        (self.0 & EXP_MASK) == 0 && (self.0 & MANTISSA_MASK) != 0
    }

    #[inline(always)]
    pub fn is_sign_negative(self) -> bool {
        (self.0 & SIGN_MASK) != 0
    }

    #[inline(always)]
    pub fn eq_value(self, other: Self) -> bool {
        if self.is_nan() || other.is_nan() {
            return false;
        }
        if self.is_zero() && other.is_zero() {
            return true;
        }
        self.0 == other.0
    }

    /// Add two `bfloat16` values (via `f32`).
    #[inline(always)]
    #[allow(clippy::should_implement_trait)]
    pub fn add(self, rhs: Self) -> Self {
        Self::from_f32(self.to_f32() + rhs.to_f32())
    }

    /// Subtract two `bfloat16` values (via `f32`).
    #[inline(always)]
    #[allow(clippy::should_implement_trait)]
    pub fn sub(self, rhs: Self) -> Self {
        Self::from_f32(self.to_f32() - rhs.to_f32())
    }

    /// Multiply two `bfloat16` values (via `f32`).
    #[inline(always)]
    #[allow(clippy::should_implement_trait)]
    pub fn mul(self, rhs: Self) -> Self {
        Self::from_f32(self.to_f32() * rhs.to_f32())
    }

    /// Divide two `bfloat16` values (via `f32`).
    #[inline(always)]
    #[allow(clippy::should_implement_trait)]
    pub fn div(self, rhs: Self) -> Self {
        Self::from_f32(self.to_f32() / rhs.to_f32())
    }

    /// Negate this `bfloat16` value.
    #[inline(always)]
    #[allow(clippy::should_implement_trait)]
    pub fn neg(self) -> Self {
        Self(self.0 ^ SIGN_MASK)
    }

    /// Absolute value.
    #[inline(always)]
    pub fn abs(self) -> Self {
        Self(self.0 & !SIGN_MASK)
    }
}

impl PartialEq for bfloat16 {
    #[inline]
    fn eq(&self, other: &Self) -> bool {
        self.0 == other.0
    }
}

impl Eq for bfloat16 {}

impl Hash for bfloat16 {
    #[inline]
    fn hash<H: Hasher>(&self, state: &mut H) {
        self.0.hash(state);
    }
}

impl PartialOrd for bfloat16 {
    #[inline]
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        self.to_f32().partial_cmp(&other.to_f32())
    }
}

impl Add for bfloat16 {
    type Output = Self;
    #[inline]
    fn add(self, rhs: Self) -> Self {
        Self::add(self, rhs)
    }
}

impl Sub for bfloat16 {
    type Output = Self;
    #[inline]
    fn sub(self, rhs: Self) -> Self {
        Self::sub(self, rhs)
    }
}

impl Mul for bfloat16 {
    type Output = Self;
    #[inline]
    fn mul(self, rhs: Self) -> Self {
        Self::mul(self, rhs)
    }
}

impl Div for bfloat16 {
    type Output = Self;
    #[inline]
    fn div(self, rhs: Self) -> Self {
        Self::div(self, rhs)
    }
}

impl Neg for bfloat16 {
    type Output = Self;
    #[inline]
    fn neg(self) -> Self {
        Self::neg(self)
    }
}

impl fmt::Display for bfloat16 {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.to_f32())
    }
}

impl fmt::Debug for bfloat16 {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "bfloat16({})", self.to_f32())
    }
}

#[cfg(test)]
mod tests {
    use super::bfloat16;

    #[test]
    fn test_basic_conversion() {
        assert_eq!(bfloat16::from_f32(1.0).to_bits(), 0x3F80);
        assert_eq!(bfloat16::from_f32(-1.0).to_bits(), 0xBF80);
        assert_eq!(bfloat16::from_f32(0.0), bfloat16::ZERO);
        assert_eq!(bfloat16::from_f32(-0.0), bfloat16::NEG_ZERO);
    }

    #[test]
    fn test_special_values() {
        assert_eq!(bfloat16::INFINITY.to_bits(), 0x7F80);
        assert_eq!(bfloat16::NEG_INFINITY.to_bits(), 0xFF80);
        assert_eq!(bfloat16::NAN.to_bits(), 0x7FC0);
        assert!(bfloat16::from_f32(f32::NAN).is_nan());
    }

    #[test]
    fn test_round_to_nearest_even() {
        assert_eq!(bfloat16::from_f32(1.0 + 1.0 / 256.0).to_bits(), 0x3F80);
        assert_eq!(bfloat16::from_f32(1.0 + 3.0 / 256.0).to_bits(), 0x3F82);
    }
}
